### ISSUE TITLE
[JENKINS-21311] Queue item cancellation via REST api works but returns a 404

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#15277]
<p>When I call http://<Jenkins_URL>/queue/cancelItem?id=<queueItem> it does cancel the item but it returns a 404. It must return a 200 instead.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=barthelemy">barthelemy</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311" target="_blank">Queue item cancellation via REST api works but returns a 404</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/pierrebtz.png" title="pierrebtz's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=pierrebtz">pierrebtz</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: core
<li><b>label(s)</b>: newbie-friendly, queue, rest, stapler
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2020-03-08T22:18:29+00:00
<li><b>votes</b>: 3
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>When I call http://&lt;Jenkins_URL&gt;/queue/cancelItem?id=&lt;queueItem&gt; it does cancel the item but it returns a 404. It must return a 200 instead.</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>jenkins 1.546</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
- _Remote link associated with this issue:_


  - [PR#​4395](https://github.com/jenkinsci/jenkins/pull/4395)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>barthelemy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=192648&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-192648">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Probably related to <a href="https://issues.jenkins.io/browse/JENKINS-16333" title="Cancel in the queue widget tends to cause 404" class="issue-link" data-issue-key="JENKINS-16333"><del>JENKINS-16333</del></a></p>
</pre>
</details>

</ul>
</details>

<p>Probably related to <a href="https://issues.jenkins.io/browse/JENKINS-16333" title="Cancel in the queue widget tends to cause 404" class="issue-link" data-issue-key="JENKINS-16333"><del>JENKINS-16333</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=213088&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213088">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is this still an issue in recent Jenkins versions? The <tt>Queue.doCancelItem</tt> code looks reasonable.</p>

<p>Where is <tt>/queue/cancelItem</tt> documented as being part of the REST API?</p>
</pre>
</details>

</ul>
</details>

<p>Is this still an issue in recent Jenkins versions? The <tt>Queue.doCancelItem</tt> code looks reasonable.</p>

<p>Where is <tt>/queue/cancelItem</tt> documented as being part of the REST API?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=213157&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213157">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is this reproducible? <tt>doCancelItem</tt> unconditionally returns a redirect response. It should never be a 404, even if the queue item does not exist.</p>
</pre>
</details>

</ul>
</details>

<p>Is this reproducible? <tt>doCancelItem</tt> unconditionally returns a redirect response. It should never be a 404, even if the queue item does not exist.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>barthelemy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=213394&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213394">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have just tried again and could reproduce it with Jenkins 1.579. </p>

<p>I did <a href="https://xxx.yyy.ch/jenkins/view/job/" class="external-link" target="_blank" rel="nofollow noopener">https://xxx.yyy.ch/jenkins/view/job/</a>&lt;jobname&gt;/buildWithParameters?module=xyz
And got a 201 with a location in the header : Location: <a href="https://xxx.yyy.ch/jenkins/queue/item/1137/" class="external-link" target="_blank" rel="nofollow noopener">https://xxx.yyy.ch/jenkins/queue/item/1137/</a></p>

<p>I then I did : <a href="https://xxx.yyy.ch/jenkins/queue/cancelItem?id=1137" class="external-link" target="_blank" rel="nofollow noopener">https://xxx.yyy.ch/jenkins/queue/cancelItem?id=1137</a>
And got the following header : 
    Status Code: 404 Not Found
    Cache-Control: no-cache
    Connection: close
    Content-Length: 952
    Content-Type: text/html;charset=utf-8
    Date: Thu, 09 Oct 2014 11:45:15 GMT
    Expires: Thu, 01 Jan 1970 01:00:00 CET
    Pragma: No-cache</p>

<p>Yet, the queue item was cancelled !</p>
</pre>
</details>

</ul>
</details>

<p>I have just tried again and could reproduce it with Jenkins 1.579. </p>

<p>I did <a href="https://xxx.yyy.ch/jenkins/view/job/" class="external-link" target="_blank" rel="nofollow noopener">https://xxx.yyy.ch/jenkins/view/job/</a><jobname>/buildWithParameters?module=xyz<br/>
And got a 201 with a location in the header : Location: <a href="https://xxx.yyy.ch/jenkins/queue/item/1137/" class="external-link" target="_blank" rel="nofollow noopener">https://xxx.yyy.ch/jenkins/queue/item/1137/</a></p>

<p>I then I did : <a href="https://xxx.yyy.ch/jenkins/queue/cancelItem?id=1137" class="external-link" target="_blank" rel="nofollow noopener">https://xxx.yyy.ch/jenkins/queue/cancelItem?id=1137</a><br/>
And got the following header : <br/>
    Status Code: 404 Not Found<br/>
    Cache-Control: no-cache<br/>
    Connection: close<br/>
    Content-Length: 952<br/>
    Content-Type: text/html;charset=utf-8<br/>
    Date: Thu, 09 Oct 2014 11:45:15 GMT<br/>
    Expires: Thu, 01 Jan 1970 01:00:00 CET<br/>
    Pragma: No-cache</p>

<p>Yet, the queue item was cancelled !</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=213399&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213399">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It makes sense to me now.</p>

<p>It redirects to the referer URL, which, if that's not set, defaults to '.', or /queue/ (which isn't mean to be used interactively, so there's nothing there &#8211; 404).</p>

<p>This is why I was asking where /queue/cancelItem was advertised as being part of the Jenkins API &#8211; it's not documented in /queue/api. If it's advertised nowhere, and you're just using it because it exists, then there's no reasonable expectation of URLs used by the Jenkins UI to also work outside that context.</p>
</pre>
</details>

</ul>
</details>

<p>It makes sense to me now.</p>

<p>It redirects to the referer URL, which, if that's not set, defaults to '.', or /queue/ (which isn't mean to be used interactively, so there's nothing there &#8211; 404).</p>

<p>This is why I was asking where /queue/cancelItem was advertised as being part of the Jenkins API &#8211; it's not documented in /queue/api. If it's advertised nowhere, and you're just using it because it exists, then there's no reasonable expectation of URLs used by the Jenkins UI to also work outside that context.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>barthelemy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=213401&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213401">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Ok, fair enough. </p>

<p>For information I got the information about it in these two places : 
<a href="http://stackoverflow.com/questions/21021905/how-to-stop-a-build-in-jenkins-via-the-rest-api" class="external-link" target="_blank" rel="nofollow noopener">http://stackoverflow.com/questions/21021905/how-to-stop-a-build-in-jenkins-via-the-rest-api</a>
<a href="https://groups.google.com/forum/#!topic/jenkinsci-users/aN_WFvgTz2Y" class="external-link" target="_blank" rel="nofollow noopener">https://groups.google.com/forum/#!topic/jenkinsci-users/aN_WFvgTz2Y</a></p>

<p>Is there a more standard way of cancelling a queue item ? </p>

</pre>
</details>

</ul>
</details>

<p>Ok, fair enough. </p>

<p>For information I got the information about it in these two places : <br/>
<a href="http://stackoverflow.com/questions/21021905/how-to-stop-a-build-in-jenkins-via-the-rest-api" class="external-link" target="_blank" rel="nofollow noopener">http://stackoverflow.com/questions/21021905/how-to-stop-a-build-in-jenkins-via-the-rest-api</a><br/>
<a href="https://groups.google.com/forum/#!topic/jenkinsci-users/aN_WFvgTz2Y" class="external-link" target="_blank" rel="nofollow noopener">https://groups.google.com/forum/#!topic/jenkinsci-users/aN_WFvgTz2Y</a></p>

<p>Is there a more standard way of cancelling a queue item ? </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>barthelemy:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=213402&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213402">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Actually the second link points to a post from you <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Actually the second link points to a post from you <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=213404&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213404">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There are lots of API endpoints which are not documented, mainly because the creator did not document them. (In this case I am at fault, though I doubt the endpoint I was replacing was documented either.)</p>

<p>So this is simply a bug. Using <tt>HttpResponses.forwardToPreviousPage()</tt> is commonplace in <tt>@RequirePOST</tt> endpoints that are not expected to display anything special where the UI invoker is just something that sends a <tt>POST</tt> via a <tt>form</tt> and does not want the rendered page to change. One easy fix would be to make this method in Stapler check whether there actually is a <tt>Referer</tt>, and if not, just delegate to <tt>ok()</tt> instead.</p>

<p>The proper fix would probably be to think of this as a real API and return 204 No Content, optionally with some body or status message indicating whether anything was actually canceled. It is also necessary to ensure that whatever status code is chosen, <tt>stopButton.jelly</tt> (called from two places with <tt>cancelItem</tt>) will not try to reload the page. From the use of <tt>onclick="new Ajax.Request(this.href); return false"</tt> it looks like this would already be true regardless of the status code.</p>

<p>And it looks like we need a <tt>Queue/_api.jelly</tt> listing the various operations. Currently this gives only generic information.</p>
</pre>
</details>

</ul>
</details>

<p>There are lots of API endpoints which are not documented, mainly because the creator did not document them. (In this case I am at fault, though I doubt the endpoint I was replacing was documented either.)</p>

<p>So this is simply a bug. Using <tt>HttpResponses.forwardToPreviousPage()</tt> is commonplace in <tt>@​RequirePOST</tt> endpoints that are not expected to display anything special where the UI invoker is just something that sends a <tt>POST</tt> via a <tt>form</tt> and does not want the rendered page to change. One easy fix would be to make this method in Stapler check whether there actually is a <tt>Referer</tt>, and if not, just delegate to <tt>ok()</tt> instead.</p>

<p>The proper fix would probably be to think of this as a real API and return 204 No Content, optionally with some body or status message indicating whether anything was actually canceled. It is also necessary to ensure that whatever status code is chosen, <tt>stopButton.jelly</tt> (called from two places with <tt>cancelItem</tt>) will not try to reload the page. From the use of <tt>onclick="new Ajax.Request(this.href); return false"</tt> it looks like this would already be true regardless of the status code.</p>

<p>And it looks like we need a <tt>Queue/_api.jelly</tt> listing the various operations. Currently this gives only generic information.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=213435&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-213435">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Alright, so it's actually a bug (although I maintain undocumented -&gt; not really API).</p>

<p>Reducing priority as it's essentially cosmetic and can be ignored; maybe just disable following the bogus redirect.</p>
</pre>
</details>

</ul>
</details>

<p>Alright, so it's actually a bug (although I maintain undocumented -> not really API).</p>

<p>Reducing priority as it's essentially cosmetic and can be ignored; maybe just disable following the bogus redirect.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/deepchip.png" title="deepchip's avatar" /> deepchip:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=335663&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-335663">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Running into this bug as I am adding the feature to cancel a queue item to the <a href="https://github.com/cdancy/jenkins-rest/pull/4" class="external-link" target="_blank" rel="nofollow noopener">jenkins-rest</a> java library. I use Jenkins 2.107.1.</p>
</pre>
</details>

</ul>
</details>

<p>Running into this bug as I am adding the feature to cancel a queue item to the <a href="https://github.com/cdancy/jenkins-rest/pull/4" class="external-link" target="_blank" rel="nofollow noopener">jenkins-rest</a> java library. I use Jenkins 2.107.1.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/pierrebtz.png" title="pierrebtz's avatar" /> pierrebtz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-21311?focusedId=381727&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381727">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I also lost some time with this return code before seeing this issue, proposed some change to address this <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>I also lost some time with this return code before seeing this issue, proposed some change to address this <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/jenkins/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-16333%22">JENKINS-16333</a>]</i>
